api/envoy/config/listener/v3/listener.proto

```diff
@@ -120,6 +121,60 @@ message Listener {
   // :ref:`FAQ entry <faq_how_to_setup_sni>`.
   repeated FilterChain filter_chains = 3;
 
+  // A filter chain matcher resolving the filter chain name based on the connection properties.
+  // The following inputs are supported by the matcher:
+  //
+  // 1. :ref:`DestinationPort <envoy_v3_api_msg_config.listener.v3.DestinationPortInput>`.
+  // 2. :ref:`DestinationIP <envoy_v3_api_msg_config.listener.v3.DestinationIPInput>`.
+  // 3. :ref:`SNI <envoy_v3_api_msg_config.listener.v3.SNIInput>`.
+  // 4. :ref:`TransportProtocol <envoy_v3_api_msg_config.listener.v3.TransportProtocolInput>`.
+  // 5. :ref:`ApplicationProtocol <envoy_v3_api_msg_config.listener.v3.ApplicationProtocolInput>`.
+  // 6. :ref:`DirectSourceIP <envoy_v3_api_msg_config.listener.v3.DirectSourceIPInput>`.
+  // 7. :ref:`SourceType <envoy_v3_api_msg_config.listener.v3.SourceTypeInput>`.
+  // 8. :ref:`SourceIP <envoy_v3_api_msg_config.listener.v3.SourceIPInput>`.
+  // 9. :ref:`SourcePort <envoy_v3_api_msg_config.listener.v3.SourcePortInput>`.
+  //
+  // If specified, all :ref:`filter_chains <envoy_v3_api_field_config.listener.v3.Listener.filter_chains>` must
+  // have a non-empty :ref:`name <envoy_v3_api_field_config.listener.v3.FilterChain.name>` and empty :ref:`filter_chain_match <envoy_v3_api_msg_config.listener.v3.FilterChain.filter_chain_match>`.
+  //
+  // Example 1: Match port 80 for chain "http", port 443 and SNI "*.server.com" for chain "server" or SNI "*.com" for chain "https".
+
+  // .. code-block:: yaml
+  //
+  //  filter_chain_matcher:
+  //    matcher_tree:
+  //      input:
+  //        typed_config:
+  //          "@type": type.googleapis.com/envoy.config.listener.v3.DestinationPortInput
+  //      exact_match_map:
+  //        map:
+  //          "80":
+  //            action:
+  //              name: http
+  //          "443":
+  //            matcher:
+  //              matcher_tree:
+  //                input:
+  //                  typed_config:
+  //                    "@type": type.googleapis.com/envoy.config.listener.v3.SNIInput
+  //                custom_match:
```

Comment:
Yes, it might be better to have a look at the [network inputs](https://github.com/envoyproxy/envoy/pull/19493) and [trie matcher](https://github.com/cncf/xds/pull/24) PRs. The inputs should generally be constant in the input, and matching should be sublinear with IP and domain names. Port ranges should use segment trees for logarithmic performance. ALPN lists are tricky to handle, which needs some core matcher framework support to improve.